// @graphiql/plugin-explorer/dist/index.mjs
import * as t from "react";
import n, { useCallback as u } from "react";
import { useEditorContext as d, useSchemaContext as m, useExecutionContext as p, useOptimisticState as v, useOperationsEditorState as g } from "@graphiql/react";
import { Explorer as w } from "graphiql-explorer";
const a = ({
  title: e,
  titleId: r,
  ...o
}) => /* @__PURE__ */ t.createElement("svg", { width: 5, height: 8, viewBox: "0 0 5 8", fill: "currentColor", xmlns: "http://www.w3.org/2000/svg", "aria-labelledby": r, ...o }, e ? /* @__PURE__ */ t.createElement("title", { id: r }, e) : null, /* @__PURE__ */ t.createElement("path", { d: "M0.910453 6.86965L3.88955 3.89061C4.09782 3.68233 4.09782 3.34465 3.88955 3.13637L0.910453 0.157278C0.574475 -0.178701 0 0.0592511 0 0.534408V6.49259C0 6.96768 0.574475 7.20565 0.910453 6.86965Z" })), x = ({
  title: e,
  titleId: r,
  ...o
}) => /* @__PURE__ */ t.createElement("svg", { height: "1em", strokeWidth: 1.5, viewBox: "0 0 24 24", stroke: "currentColor", fill: "none", xmlns: "http://www.w3.org/2000/svg", "aria-labelledby": r, ...o }, e ? /* @__PURE__ */ t.createElement("title", { id: r }, e) : null, /* @__PURE__ */ t.createElement("path", { d: "M18 6H20M22 6H20M20 6V4M20 6V8", strokeLinecap: "round", strokeLinejoin: "round" }), /* @__PURE__ */ t.createElement("path", { d: "M21.4 20H2.6C2.26863 20 2 19.7314 2 19.4V11H21.4C21.7314 11 22 11.2686 22 11.6V19.4C22 19.7314 21.7314 20 21.4 20Z", strokeLinecap: "round", strokeLinejoin: "round" }), /* @__PURE__ */ t.createElement("path", { d: "M2 11V4.6C2 4.26863 2.26863 4 2.6 4H8.77805C8.92127 4 9.05977 4.05124 9.16852 4.14445L12.3315 6.85555C12.4402 6.94876 12.5787 7 12.722 7H14", strokeLinecap: "round", strokeLinejoin: "round" })), E = ({
  title: e,
  titleId: r,
  ...o
}) => /* @__PURE__ */ t.createElement("svg", { width: 15, height: 15, viewBox: "0 0 15 15", xmlns: "http://www.w3.org/2000/svg", stroke: "currentColor", fill: "none", "aria-labelledby": r, ...o }, e ? /* @__PURE__ */ t.createElement("title", { id: r }, e) : null, /* @__PURE__ */ t.createElement("circle", { cx: 7.5, cy: 7.5, r: 6, strokeWidth: 2 })), y = ({
  title: e,
  titleId: r,
  ...o
}) => /* @__PURE__ */ t.createElement("svg", { width: 15, height: 15, viewBox: "0 0 15 15", xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", "aria-labelledby": r, ...o }, e ? /* @__PURE__ */ t.createElement("title", { id: r }, e) : null, /* @__PURE__ */ t.createElement("circle", { cx: 7.5, cy: 7.5, r: 7.5 }), /* @__PURE__ */ t.createElement("path", { d: "M4.64641 7.00106L6.8801 9.23256L10.5017 5.61325", stroke: "white", strokeWidth: 1.5 }));
const k = {
  keyword: "hsl(var(--color-primary))",
  def: "hsl(var(--color-tertiary))",
  property: "hsl(var(--color-info))",
  qualifier: "hsl(var(--color-secondary))",
  attribute: "hsl(var(--color-tertiary))",
  number: "hsl(var(--color-success))",
  string: "hsl(var(--color-warning))",
  builtin: "hsl(var(--color-success))",
  string2: "hsl(var(--color-secondary))",
  variable: "hsl(var(--color-secondary))",
  atom: "hsl(var(--color-tertiary))"
}, C = /* @__PURE__ */ n.createElement(a, { style: { width: "var(--px-16)", transform: "rotate(90deg)" } }), f = /* @__PURE__ */ n.createElement(a, { style: { width: "var(--px-16)" } }), b = /* @__PURE__ */ n.createElement(E, { style: { marginRight: "var(--px-4)" } }), L = /* @__PURE__ */ n.createElement(
  y,
  {
    style: { fill: "hsl(var(--color-info))", marginRight: "var(--px-4)" }
  }
), S = {
  buttonStyle: {
    cursor: "pointer",
    fontSize: "2em",
    lineHeight: 0
  },
  explorerActionsStyle: {
    paddingTop: "var(--px-16)"
  },
  actionButtonStyle: {}
};
function M(e) {
  const { setOperationName: r } = d({ nonNull: !0 }), { schema: o } = m({ nonNull: !0 }), { run: l } = p({ nonNull: !0 }), i = u(
    (c) => {
      c && r(c), l();
    },
    [l, r]
  ), [s, h] = v(
    g()
  );
  return /* @__PURE__ */ n.createElement(
    w,
    {
      schema: o,
      onRunOperation: i,
      explorerIsOpen: !0,
      colors: k,
      arrowOpen: C,
      arrowClosed: f,
      checkboxUnchecked: b,
      checkboxChecked: L,
      styles: S,
      query: s,
      onEdit: h,
      ...e
    }
  );
}
function V(e) {
  return {
    title: "GraphiQL Explorer",
    icon: x,
    content: () => /* @__PURE__ */ n.createElement(M, { ...e })
  };
}
export {
  V as explorerPlugin
};
